--- a/net/minecraft/server/commands/CommandSummon.java
+++ b/net/minecraft/server/commands/CommandSummon.java
@@ -25,6 +25,14 @@
 import net.minecraft.world.level.World;
 import net.minecraft.world.phys.Vec3D;
 
+// CraftBukkit start
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.world.entity.EntityLiving;
+import net.minecraft.world.item.ItemStack;
+import org.bukkit.Bukkit;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
+// CraftBukkit end
+
 public class CommandSummon {
 
     private static final SimpleCommandExceptionType ERROR_FAILED = new SimpleCommandExceptionType(IChatBaseComponent.translatable("commands.summon.failed"));
@@ -67,7 +75,21 @@
                     ((EntityInsentient) entity).finalizeSpawn(commandlistenerwrapper.getLevel(), commandlistenerwrapper.getLevel().getCurrentDifficultyAt(entity.blockPosition()), EnumMobSpawn.COMMAND, (GroupDataEntity) null);
                 }
 
-                if (!worldserver.tryAddFreshEntityWithPassengers(entity)) {
+                Bukkit.broadcastMessage("1 " + (entity instanceof EntityLiving) + " " + nbttagcompound.contains("ArmorItems", 9));
+                // CraftBukkit start
+                if (entity instanceof EntityLiving entityLiving && nbttagcompound.contains("ArmorItems", 9)) {
+                    Bukkit.broadcastMessage("2");
+                    NBTTagList nbtTagList = nbttagcompound.getList("ArmorItems", 10);
+                    for (int i = 0; i < nbtTagList.size(); i++) {
+                        ItemStack itemStack = ItemStack.parseOptional(entityLiving.registryAccess(), nbtTagList.getCompound(i));
+                        Bukkit.broadcastMessage("3 " + itemStack);
+                        if (!ItemStack.isSameItem(itemStack, ItemStack.EMPTY)) {
+                            CraftEventFactory.callEntityEquipArmorEvent(entityLiving, ItemStack.EMPTY, itemStack, 8 - i);
+                        }
+                    }
+                }
+                // CraftBukkit end
+                if (!worldserver.tryAddFreshEntityWithPassengers(entity, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.COMMAND)) { // CraftBukkit - pass a spawn reason of "COMMAND"
                     throw CommandSummon.ERROR_DUPLICATE_UUID.create();
                 } else {
                     return entity;
